runtime/array.c

```diff
@@ -303,24 +306,112 @@ CAMLprim value caml_make_array(value init)
 
 /* Blitting */
 
+/* [wo_memmove] copies [nvals] values from [src] to [dst]. If there is a single
+   domain running, then we use [memmove]. Otherwise, we copy one word at a
+   time.
+
+   Since the [memmove] implementation does not guarantee that the writes are
+   always word-sized, we explicitly perform word-sized writes of the release
+   kind to avoid mixed-mode accesses. Performing release writes should be
+   sufficient to prevent smart compilers from coalesing the writes into vector
+   writes, and hence prevent mixed-mode accesses. [MM].
+   */
+static void wo_memmove (value* const dst, const value* const src,
+                        mlsize_t nvals)
+{
+  mlsize_t i;
+
+  if (caml_domain_alone ()) {
+    memmove (dst, src, nvals * sizeof (value));
+  } else {
+    /* See memory model [MM] notes in memory.c */
+    atomic_thread_fence(memory_order_acquire);
+    if (dst < src) {
+      /* copy ascending */
+      for (i = 0; i < nvals; i++)
+        atomic_store_explicit(&((atomic_value*)dst)[i], src[i],
+                              memory_order_release);
+
+    } else {
+      /* copy descending */
+      for (i = nvals; i > 0; i--)
+        atomic_store_explicit(&((atomic_value*)dst)[i-1], src[i-1],
+                              memory_order_release);
+    }
+  }
+}
+
+/* See comments attached to [wo_memmove] */
 CAMLprim value caml_floatarray_blit(value a1, value ofs1, value a2, value ofs2,
                                     value n)
 {
-  memmove((double *)a2 + Long_val(ofs2),
-          (double *)a1 + Long_val(ofs1),
-          Long_val(n) * sizeof(double));
+  double *dst, *src;
+  mlsize_t nvals, i;
+
+  dst = (double *)a2 + Long_val(ofs2);
+  src = (double *)a1 + Long_val(ofs1);
+  nvals = Long_val(n);
+
+  if (caml_domain_alone ()) {
+    memmove (dst, src, nvals * sizeof(double));
+  } else {
+    /* See memory model [MM] notes in memory.c */
+    atomic_thread_fence(memory_order_acquire);
+    if (dst < src) {
+      /* copy ascending */
+      for (i = 0; i < nvals; i++)
+        atomic_store_explicit(&((atomic_double*)dst)[i], src[i],
+                              memory_order_release);
```

Comment:
Apologies for noticing after this PR was merged, but: this use of `atomic_double` does not work on ARM 32 bits and may not work on other platforms.  On ARM 32 bits, GCC implements `_Atomic double` as a double protected by a mutex.  So, it doesn't have the memory layout of a `double`, and atomic operations call into library functions from the `-latomic` library, which call  POSIX thread mutex operations.

What this means is that the OCaml runtime will not even compile on ARM32 in bytecode-only mode, and if we add `-latomic` the blit operation over float arrays will crash or hang or write garbage.

I think it would be  preferable to leave open the memory model issues involving non-word-sized accesses and discuss them thoroughly later rather than trying quick "fixes" that don't quite work.

Reply:
Thanks for noticing this. I agree with not inserting quick fixes for memory model. 

How should we proceed with this now? I can remove `atomic_double` and revert `caml_floatarray_blit` to trunk version with a comment about the mixed-mode accesses? It may be useful to retain the fence before `memmove` since that would anyway be needed to enforce the memory model.

Reply:
For the time being, maybe you could keep the `memmove`-based implementation and just add a comment saying "not consistent with the memory model, to be discussed later".

Later, if we agree to keep supporting 32-bit platforms but only in single-domain mode, we could use `wo_memmove` for 64-bit platforms and `memmove` for 32-bit platforms.  But maybe we could finally agree to drop all support for 32-bit platforms...

Reply:
To clarify, it is only the `caml_floatarray_blit` that is problematic. `wo_memmove` is not used on float arrays, and IINM is safe on 32-bit platforms. Please correct me if I am wrong. 

> For the time being, maybe you could keep the memmove-based implementation and just add a comment saying "not consistent with the memory model, to be discussed later".

With that, are you suggesting that I revert the `caml_floatarray_blit` implementation to trunk version and add the suggested comment?

Reply:
> With that, are you suggesting that I revert the caml_floatarray_blit implementation to trunk version and add the suggested comment?

Yes, that's my suggestion, but feel free to disagree :-) 

My reasoning is that we need to discuss more globally the memory model for byte arrays, flat float arrays, and bigarrays, and agree on what we want before writing code.

Reply:
I’ve opened https://github.com/ocaml-multicore/ocaml-multicore/pull/831.